### PR TITLE
Authentication Revamp + WIP Referral UI

### PR DESCRIPTION
**Pull request by jeevansurendran**

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    package="com.checkin.app.checkin">
+    package="com.checkin.app.checkin"> <!-- Fixes Manifest Merge error with Razorpay library -->
+    <uses-sdk tools:overrideLibrary="com.razorpay" />
```
Comment (shivanshs9) [inline app/src/main/AndroidManifest.xml:5]: Why are these lines moved?

File: app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt [hunk 2]
```diff
@@ -17,12 +17,14 @@ class AuthViewModel(application: Application) : BaseViewModel(application) {
...
 
-    private var firebaseIdToken: String? = null
+    var firebaseIdToken: String? = null
+        private set
     private var providerIdToken: String? = null
 
     var isLoginAttempt = false
         private set
-
+    var isRegisteredAttempt = false
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt:26]: Doesn't `isRegisteredAttempt = !isLoginAttempt` always?

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+class AuthenticationActivity : AppCompatActivity(), AuthFragmentInteraction, AuthOtpFragment.AuthCallback {
+    @BindView(R.id.fl_circle_progress_container)
+    internal lateinit var flLoadingContainer: FrameLayout
+    @BindView(R.id.tv_auth_terms_conditions)
+    internal lateinit var tvTermsAndConditions: TextView
+
+    private val firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()
+    private val authViewModel: AuthViewModel by viewModels()
+    private val navController: NavController get() = findNavController(R.id.nav_host_authentication)
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:56]: Use lazy instead of overriding get()
Also, get() should always be in a new line

File: app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        }
+
+    init {
+        View.inflate(context, R.layout.view_otp, this).apply {
+            ButterKnife.bind(this@OtpView)
+        }
+
+        //adding
+        etArray[0].addTextChangedListener(GenericTextWatcher(etArray[0], etArray[1]))
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt:54]: It would be neat to use a loop on **etArray**

File: app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        etArray[3].addTextChangedListener(GenericTextWatcher(etArray[3], etArray[4]))
+        etArray[4].addTextChangedListener(GenericTextWatcher(etArray[4], etArray[5]))
+        etArray[5].addTextChangedListener(GenericTextWatcher(etArray[5], null))
+
+        //deleting
+        etArray[0].setOnKeyListener(GenericKeyEvent(etArray[0], null))
+        etArray[1].setOnKeyListener(GenericKeyEvent(etArray[1], etArray[0]))
+        etArray[2].setOnKeyListener(GenericKeyEvent(etArray[2], etArray[1]))
+        etArray[3].setOnKeyListener(GenericKeyEvent(etArray[3], etArray[2]))
+        etArray[4].setOnKeyListener(GenericKeyEvent(etArray[4], etArray[3]))
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt:66]: Same here

File: app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+
+}
+
+class GenericTextWatcher internal constructor(private val currentView: View, private val nextView: View?) : TextWatcher {
+    override fun afterTextChanged(editable: Editable) {
+        val text = editable.toString()
+        when (currentView.id) {
+            R.id.otp1 -> if (text.length == 1) nextView!!.requestFocus()
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/misc/views/OtpView.kt:103]: All the code is same, just **R.id.otp6** case is different. But in that case, **nextView** is null anyways so you could rewrite the logic to 1 line only

File: app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt (hunk 2 above)
Comment (jeevansurendran) [reply app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt:26]: not always if login fails, still login attempt is  true

Comment (jeevansurendran) [reply app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt:26]: and you need `isLoginAttempt` to be `true` while login fails because when we get the result we use this condition to check if actually a login attempt is made and it as failed
`
Resource.Status.ERROR_INVALID_REQUEST -> if (authViewModel.isLoginAttempt) {
                        if (!::phoneNo.isInitialized) {
                            toast("Invalid Mobile Number")
                            pass
                        }
                        askUserDetails()
                    }
`

Comment (jeevansurendran) [reply app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt:26]: 'isRegisteredAttempt = !isLoginAttempt' but i can remove isRegistedAttempt too and i will do that

Comment (shivanshs9) [reply app/src/main/java/com/checkin/app/checkin/auth/AuthViewModel.kt:26]: Yeah I think it's possible to do without that

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        }
+        val name = user.displayName
+        goBack = false
+
+        val fullNameSeparatedList = name?.split("\\s+".toRegex()) ?: listOf(defaultFirstName)
+        //create an user
+        onUserInfoProcess(firstName = fullNameSeparatedList[0], lastName = if (fullNameSeparatedList.size > 1) fullNameSeparatedList[1]
+        else null, username = phoneNo, gender = defaultGender)
+
+        authViewModel.authResult.observe(this, Observer {
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:150]: I'm not sure why you're trying to add another observer here... since all the observers would be called and you're adding it everytime on call of _askUserDetails_

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                            }
+                        }
+                    }
+                }
+            })
+            userViewModel.fetchUserData()
+        }
+    }
+
+    private fun openHome() {
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:230]: Instead of creating a new function, reuse an [existing one](https://github.com/checkin-team/Checkin-Android/blob/3454cf0730aeedcceecadae5e66c0133175e323e/app/src/main/java/com/checkin/app/checkin/utility/Utils.java#L117)

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+    }
+
+    override fun onCancelVerification() {
+
+    }
+
+    override fun onFailedVerification(exception: Exception) {
+        toast(exception.message ?: getString(R.string.error_authentication_phone))
+        if (exception !is InvalidOTPException) {
+            toast("Invalid OTP try Again")
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:249]: you're showing two toast one after another here... bad UX
Use proper if-else and also log exceptions (Use Utils.log for it)

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
+                try {
+                    val account = task.getResult(ApiException::class.java)
+                    authViewModel.setProviderIdToken(account!!.idToken)
+                    val credential = GoogleAuthProvider.getCredential(account.idToken, null)
+                    authenticateWithCredential(credential)
+                } catch (e: ApiException) {
+                    Log.e(TAG, "GoogleAuth - Verification Failed: ", e)
+                    hideProgress()
+                    Toast.makeText(applicationContext, R.string.error_authentication_google, Toast.LENGTH_SHORT).show()
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:276]: use toast extension function

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        when (requestCode) {
+            RC_AUTH_GOOGLE -> {
+                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
+                try {
+                    val account = task.getResult(ApiException::class.java)
+                    authViewModel.setProviderIdToken(account!!.idToken)
+                    val credential = GoogleAuthProvider.getCredential(account.idToken, null)
+                    authenticateWithCredential(credential)
+                } catch (e: ApiException) {
+                    Log.e(TAG, "GoogleAuth - Verification Failed: ", e)
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:274]: Use [Utils.logErrors](https://github.com/checkin-team/Checkin-Android/blob/3454cf0730aeedcceecadae5e66c0133175e323e/app/src/main/java/com/checkin/app/checkin/utility/Utils.java#L534)

File: app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+        toast(exception.message ?: getString(R.string.error_authentication_phone))
+        if (exception !is InvalidOTPException) {
+            toast("Invalid OTP try Again")
+        }
+    }
+
+    override fun onBackPressed() {
+        if (goBack) {
+            super.onBackPressed()
+        }
```
Comment (shivanshs9) [inline app/src/main/java/com/checkin/app/checkin/auth/activities/AuthenticationActivity.kt:256]: for good UX, show some toast explaining why user can't back out
